src/realm/util/from_chars.hpp

```diff
@@ -0,0 +1,78 @@
+#ifndef FROM_CHARS_HPP
+#define FROM_CHARS_HPP
+#include <algorithm>
+#include <system_error>
+#include <type_traits>
+
+#include <realm/util/safe_int_ops.hpp>
+
+namespace realm::util {
+struct FromCharsResult {
+    const char* ptr;
+    std::errc ec;
+};
+
+/*
+ * from_chars implements the interface and behavior of https://en.cppreference.com/w/cpp/utility/from_chars
+ * for integer values. When from_chars is implemented on all supported platforms/compilers, we can remove
+ * this implementation.
+ *
+ * Use this if you need to parse a string into an integer value without any allocations or having to worry
+ * about any user-set locale.
+ */
+template <typename T>
+FromCharsResult from_chars(const char* const first, const char* const last, T& value, int base = 10) noexcept
+{
+    const auto digit_value = [](char c) -> uint8_t {
+        if (c >= '0' && c <= '9')
+            return uint8_t(c - '0');
+        if (c >= 'a' && c <= 'z')
+            return uint8_t(c - 'a' + 10);
+        if (c >= 'A' && c <= 'Z')
+            return uint8_t(c - 'A' + 10);
+        return 36; // Illegal digit value for all supported bases.
+    };
+
+    if (first == last) {
+        return {first, std::errc::invalid_argument};
+    }
+
+    auto ptr = first;
+    if constexpr (std::is_signed_v<T>) {
+        if (*ptr == '-') {
+            ++ptr;
+        }
+    }
+
+    T res_value = 0;
+    for (; ptr != last; ++ptr) {
+        auto digit = digit_value(*ptr);
+        if (digit >= base) {
+            break;
+        }
+
+        if (int_multiply_with_overflow_detect(res_value, base) || int_add_with_overflow_detect(res_value, digit)) {
```

Comment:
A matter of taste, but I would probably dumb this down to just

```c++
bool overflow = int_multiply_with_overflow_detect(res_value, base);
if (!overflow) {
    overflow = int_add_with_overflow_detect(res_value, digit);
}
if (overflow) {
     // ...
}
```

Reply:
Sure.